playbooks/phobos_vpn.yml

```diff
@@ -1,35 +1,95 @@
+      # NOTE(coreywright): The Phobos VPN server cannot be connected to from
+      # within the Phobos environment, so use this task as a method of
+      # determining whether we are in the Phobos environment or not.
+      command: netcat -v -z -u {{ gateway }} 4500
+      ignore_errors: true
+      changed_when: false
+      register: ping_vpn
 
-    - name: Template vpnc config
-      template:
-        src: vpnc.conf.j2
-        dest: /etc/vpnc/phobos.conf
+    - name: Configure and start VPN client
+      # NOTE(coreywright): If we are not in the Phobos environment, ie we can
+      # ping the Phobos VPN server, then we need to connect to it by way of
+      # VPN.
+      when: ping_vpn | succeeded
+      block:
+        # Use shell because python-apt isn't available so the apt module fails.
+        # I don't want to add python-apt to one of the more general slave setup
+        # scripts as it depends on python. We want to reduce the number of deps
+        # installed on general slaves to the bare minimum.
+        - name: Install apt packages
+          shell: |
+            apt-get update
+            apt-get install -y vpnc
+          register: install_packages
+          until: install_packages | success
+          retries: 5
+          delay: 2
+          tags:
+            - skip_ansible_lint
 
-    - name: Template vpnc systemd service
-      template:
-        src: vpnc.service.j2
-        dest: /lib/systemd/system/vpnc.service
 
-    - name: Start vpnc service
-      systemd:
-        enabled: yes
-        service: vpnc.service
-        daemon_reload: yes
-        state: restarted
+        - name: Template vpnc config
+          template:
+            src: vpnc.conf.j2
+            dest: /etc/vpnc/phobos.conf
+
+        - name: Template vpnc systemd service
+          template:
+            src: vpnc.service.j2
+            dest: /lib/systemd/system/vpnc.service
+
+        - name: Start VPN
+          block:
+            - name: Start vpnc service
+              systemd:
+                enabled: yes
+                service: vpnc.service
+                daemon_reload: yes
+                state: restarted
+              register: vpnc_restart
+
+          rescue:
+            - name: Output failed task information
+              debug: var=vpnc_restart
+
+            - name: Get vpnc service status
+              command: systemctl status vpnc.service
+              register: vpnc_status
+              failed_when: false
+              changed_when: false
+
+            - name: Output vpnc service status
+              debug: var=vpnc_status
+
+            - name: Get system log
+              command: journalctl -xe
+              register: system_log
+              changed_when: false
+
+            - name: Output system log
+              debug: var=system_log
+
+            - name: Exit with failure
+              fail: msg="vpnc service failed to start"
+
+    - name: Test connectivity to Phobos API endpoint
```

Comment:
wait_for can also be used here rather than netcat

Reply:
yeah, i'll use `wait_for` here.  netcat isn't working in the UDP-for-VPN case, so i'm reverting to ping there and `wait_for` suffices for TCP-for-API here.

thanks for the `wait_for` reminder (as i've primarily only used it to detect if a host is up and ready to continue with Ansible after a reboot by `wait_for` TCP port 22, ie SSH, which is what Ansible specifically needs to "wait for" before continuing).

Reply:
Assuming this endpoint is a web service, you can also use the ``uri`` module instead, which allows you to just get the header and you can tell it what HTTP response to expect.